File: app/MeetingTranscriber/ClassScribeSources/CaptureSignalHealth.swift
import Foundation

/// Content/transport health of one capture attempt. This is deliberately not
/// a lifecycle phase and it never claims that an audible signal is speech.
enum CaptureSignalState: String, Codable, CaseIterable, Sendable {
    case awaitingCallbacks
    case noCallbacks
    case silent
    case audible
}

/// Budgets and the signal floor are kept together so a platform adapter does
/// not grow unrelated magic numbers. The silence floor is an energy advisory,
/// not a VAD or voice classifier.
struct CaptureSignalThresholds: Equatable, Sendable {
    let initialCallbackBudget: TimeInterval
    let stallBudget: TimeInterval
    let silenceEnergyDBFS: Double

    init(
        initialCallbackBudget: TimeInterval,
        stallBudget: TimeInterval,
        silenceEnergyDBFS: Double,
    ) {
        self.initialCallbackBudget = max(0, initialCallbackBudget)
        self.stallBudget = max(0, stallBudget)
        self.silenceEnergyDBFS = silenceEnergyDBFS.isFinite ? silenceEnergyDBFS : -120
    }

    /// Existing macOS application-capture budgets: the CATap registration
    /// gate is three seconds, while the first callback has a larger budget for
    /// route negotiation. The -43 dBFS floor is the existing live-audio
    /// silence floor used for paragraph analysis; it remains provisional until
    /// a physical calibration corpus exists.
    static let macOSApplication = Self(
        initialCallbackBudget: 30,
        stallBudget: 12,
        silenceEnergyDBFS: -43,
    )

    /// The legacy microphone route has a shorter first-callback budget. It is
    /// still evaluated by the same signal model without moving its native
    /// setup/teardown to CaptureNativeExecutor.
    static let macOSMicrophone = Self(
        initialCallbackBudget: 2.5,
        stallBudget: 12,
        silenceEnergyDBFS: -43,
    )

    static let test = Self(
        initialCallbackBudget: 5,
        stallBudget: 3,
        silenceEnergyDBFS: -40,
    )
}

/// Clock abstraction for all signal-health deadlines. Production uses
/// systemUptime; tests provide a deterministic fake without sleeping.
protocol CaptureMonotonicClock: Sendable {
    func now() -> TimeInterval
}

struct SystemCaptureMonotonicClock: CaptureMonotonicClock {
    func now() -> TimeInterval {
        ProcessInfo.processInfo.systemUptime
    }
}

/// Per-callback evidence reduced on the adapter boundary. A zero-length
/// callback is still evidence that the transport is alive; its energy is zero.
struct CaptureSignalMeasurement: Equatable, Sendable {
    let sampleCount: Int64
    let frameCount: Int64
    let rms: Double
    let energyDBFS: Double

    init(sampleCount: Int64, frameCount: Int64, rms: Double) {
        self.sampleCount = max(0, sampleCount)
        self.frameCount = max(0, frameCount)
        let safeRMS = rms.isFinite ? max(0, rms) : 0
        self.rms = safeRMS
        self.energyDBFS = safeRMS > 0
            ? max(-120, 20 * log10(safeRMS))
            : -120
    }

    init(samples: [Float], channelCount: Int) {
        let count = Int64(samples.count)
        guard channelCount > 0, !samples.isEmpty else {
            self.init(sampleCount: count, frameCount: 0, rms: 0)
            return
        }

        var sumOfSquares = 0.0
        for sample in samples {
            let value = Double(sample)
            guard value.isFinite else { continue }
            sumOfSquares += value * value
        }
        let rms = sumOfSquares > 0
            ? sqrt(sumOfSquares / Double(samples.count))
            : 0
        self.init(
            sampleCount: count,
            frameCount: count / Int64(channelCount),
            rms: rms,
        )
    }
}

struct CaptureSignalHealthSnapshot: Equatable, Sendable {
    let attempt: SessionAttemptID
    let state: CaptureSignalState
    let startedAtMonotonic: TimeInterval
    let lastCallbackAtMonotonic: TimeInterval?
    let callbackCount: Int64
    let sampleCount: Int64
    let frameCount: Int64
    let rms: Double
    let energyDBFS: Double
    let elapsedSinceStart: TimeInterval
    let elapsedSinceLastCallback: TimeInterval?

    var hasReceivedCallbacks: Bool {
        callbackCount > 0
    }

    /// `silent` is healthy transport; only no callbacks is a capture-health
    /// failure candidate. Audible intentionally does not mean voice/speech.
    var transportIsHealthy: Bool {
        switch state {
        case .awaitingCallbacks, .noCallbacks: false
        case .silent, .audible: true
        }
    }
}

/// Lock-protected health tracker owned by exactly one active attempt. The
/// audio callback only performs a short measurement/update; it never waits on
/// an actor, MainActor, file, timer, or native resource.
final class CaptureSignalHealthTracker: @unchecked Sendable {
    private let lock = NSLock()
    private let thresholds: CaptureSignalThresholds
    private let clock: any CaptureMonotonicClock
    private var activeAttempt: SessionAttemptID?
    private var startedAtMonotonic = 0.0
    private var lastCallbackAtMonotonic: TimeInterval?
    private var callbackCount: Int64 = 0
    private var sampleCount: Int64 = 0
    private var frameCount: Int64 = 0
    private var lastRMS = 0.0
    private var lastEnergyDBFS = -120.0

    init(
        attempt: SessionAttemptID,
        thresholds: CaptureSignalThresholds,
        clock: any CaptureMonotonicClock = SystemCaptureMonotonicClock(),
    ) {
        self.thresholds = thresholds
        self.clock = clock
        begin(attempt)
    }

    func begin(_ attempt: SessionAttemptID, at now: TimeInterval? = nil) {
        let start = now ?? clock.now()
        lock.lock()
        activeAttempt = attempt
        startedAtMonotonic = start
        lastCallbackAtMonotonic = nil
        callbackCount = 0
        sampleCount = 0
        frameCount = 0
        lastRMS = 0
        lastEnergyDBFS = -120
        lock.unlock()
    }

    func invalidate(_ attempt: SessionAttemptID) {
        lock.lock()
        if activeAttempt == attempt {
            activeAttempt = nil
        }
        lock.unlock()
    }

    /// Returns false for a stale callback. The ownership check and the update
    /// are one critical section, so A cannot update B around invalidate/begin.
    @discardableResult
    func recordCallback(
        for attempt: SessionAttemptID,
        measurement: CaptureSignalMeasurement,
        at now: TimeInterval? = nil,
    ) -> Bool {
        let callbackTime = now ?? clock.now()
        lock.lock()
        guard activeAttempt == attempt else {
            lock.unlock()
            return false
        }
        callbackCount &+= 1
        sampleCount &+= measurement.sampleCount
        frameCount &+= measurement.frameCount
        lastCallbackAtMonotonic = callbackTime
        lastRMS = measurement.rms
        lastEnergyDBFS = measurement.energyDBFS
        lock.unlock()
        return true
    }

    func snapshot(
        for expectedAttempt: SessionAttemptID? = nil,
        at now: TimeInterval? = nil,
    ) -> CaptureSignalHealthSnapshot? {
        let currentTime = now ?? clock.now()
        lock.lock()
        guard let activeAttempt,
              expectedAttempt == nil || expectedAttempt == activeAttempt
        else {
            lock.unlock()
            return nil
        }

        let elapsedSinceStart = max(0, currentTime - startedAtMonotonic)
        let elapsedSinceLastCallback = lastCallbackAtMonotonic.map {
            max(0, currentTime - $0)
        }
        let state: CaptureSignalState
        if callbackCount == 0 {
            state = elapsedSinceStart >= thresholds.initialCallbackBudget
                ? .noCallbacks
                : .awaitingCallbacks
        } else if let elapsedSinceLastCallback,
                  elapsedSinceLastCallback >= thresholds.stallBudget {
            state = .noCallbacks
        } else {
            state = lastEnergyDBFS <= thresholds.silenceEnergyDBFS ? .silent : .audible
        }

        let snapshot = CaptureSignalHealthSnapshot(
            attempt: activeAttempt,
            state: state,
            startedAtMonotonic: startedAtMonotonic,
            lastCallbackAtMonotonic: lastCallbackAtMonotonic,
            callbackCount: callbackCount,
            sampleCount: sampleCount,
            frameCount: frameCount,
            rms: lastRMS,
            energyDBFS: lastEnergyDBFS,
            elapsedSinceStart: elapsedSinceStart,
            elapsedSinceLastCallback: elapsedSinceLastCallback,
        )
        lock.unlock()
        return snapshot
    }
}

File: app/ClassScribe.Windows/src/ClassScribe.Core/CaptureSignalHealth.cs
using System.Buffers.Binary;
using System.Diagnostics;

namespace ClassScribe.Core;

/// Transport/content health is separate from CapturePhase, AsrPhase, and
/// SessionPhase. Audible only means material energy; it is never a voice/VAD
/// claim.
public enum CaptureSignalState
{
    AwaitingCallbacks,
    NoCallbacks,
    Silent,
    Audible,
}

public sealed record CaptureSignalThresholds(
    double InitialCallbackBudgetSeconds,
    double StallBudgetSeconds,
    double SilenceEnergyDbfs)
{
    public static CaptureSignalThresholds Windows { get; } = new(10, 12, -43);

    public static CaptureSignalThresholds Test { get; } = new(5, 3, -40);
}

public interface IMonotonicClock
{
    double NowSeconds { get; }
}

public sealed class StopwatchMonotonicClock : IMonotonicClock
{
    public double NowSeconds => Stopwatch.GetTimestamp() / (double)Stopwatch.Frequency;
}

public readonly record struct CaptureSignalMeasurement(
    long SampleCount,
    long FrameCount,
    double Rms,
    double EnergyDbfs)
{
    public static CaptureSignalMeasurement FromPcm16(ReadOnlySpan<byte> pcm, int channels)
    {
        var sampleCount = pcm.Length / sizeof(short);
        if (sampleCount == 0)
        {
            return new CaptureSignalMeasurement(0, 0, 0, -120);
        }

        if (channels <= 0)
        {
            return new CaptureSignalMeasurement(sampleCount, 0, 0, -120);
        }

        double sumOfSquares = 0;
        for (var index = 0; index < sampleCount; index++)
        {
            var sample = BinaryPrimitives.ReadInt16LittleEndian(pcm.Slice(index * sizeof(short), sizeof(short)));
            var normalized = sample / 32768d;
            sumOfSquares += normalized * normalized;
        }

        var rms = Math.Sqrt(sumOfSquares / sampleCount);
        return FromRms(sampleCount, sampleCount / channels, rms);
    }

    public static CaptureSignalMeasurement FromPcm(
        ReadOnlySpan<byte> pcm,
        AudioPcmFormat format)
    {
        format.EnsureValid();
        var samples = PcmAudioConverter.Decode(pcm, format);
        if (samples.Length == 0)
        {
            return new CaptureSignalMeasurement(0, 0, 0, -120);
        }

        double sumOfSquares = 0;
        foreach (var sample in samples)
        {
            var normalized = float.IsFinite(sample) ? sample : 0;
            sumOfSquares += normalized * normalized;
        }

        return FromRms(
            samples.Length,
            samples.Length / format.Channels,
            Math.Sqrt(sumOfSquares / samples.Length));
    }

    public static CaptureSignalMeasurement FromRms(long sampleCount, long frameCount, double rms)
    {
        var safeRms = double.IsFinite(rms) ? Math.Max(0, rms) : 0;
        var energyDbfs = safeRms > 0
            ? Math.Max(-120, 20 * Math.Log10(safeRms))
            : -120;
        return new CaptureSignalMeasurement(
            Math.Max(0, sampleCount),
            Math.Max(0, frameCount),
            safeRms,
            energyDbfs);
    }
}

public sealed record CaptureSignalHealthSnapshot(
    SessionAttemptID Attempt,
    CaptureSignalState State,
    double StartedAtMonotonic,
    double? LastCallbackAtMonotonic,
    long CallbackCount,
    long SampleCount,
    long FrameCount,
    double Rms,
    double EnergyDbfs,
    double ElapsedSinceStart,
    double? ElapsedSinceLastCallback)
{
    public bool HasReceivedCallbacks => CallbackCount > 0;
}

/// Separates observing a structured signal state from presenting its status.
/// A state observed during startup is still presentable once recording begins.
public static class CaptureSignalPresentationPolicy
{
    public static bool ShouldPresent(
        CaptureSignalState state,
        CaptureSignalState? lastPresentedState,
        bool isRecording) => isRecording && state != lastPresentedState;
}

/// Thread-safe per-attempt classifier used by the WASAPI callback adapter.
/// Ownership and state update share one lock so stale A cannot update B.
public sealed class CaptureSignalHealthTracker
{
    private readonly object sync = new();
    private readonly CaptureSignalThresholds thresholds;
    private readonly IMonotonicClock clock;
    private SessionAttemptID? activeAttempt;
    private double startedAtMonotonic;
    private double? lastCallbackAtMonotonic;
    private long callbackCount;
    private long sampleCount;
    private long frameCount;
    private double lastRms;
    private double lastEnergyDbfs = -120;

    public CaptureSignalHealthTracker(
        CaptureSignalThresholds thresholds,
        IMonotonicClock? clock = null)
    {
        this.thresholds = thresholds ?? throw new ArgumentNullException(nameof(thresholds));
        this.clock = clock ?? new StopwatchMonotonicClock();
    }

    public void Begin(SessionAttemptID attempt, double? nowSeconds = null)
    {
        ArgumentNullException.ThrowIfNull(attempt);
        var start = nowSeconds ?? clock.NowSeconds;
        lock (sync)
        {
            activeAttempt = attempt;
            startedAtMonotonic = start;
            lastCallbackAtMonotonic = null;
            callbackCount = 0;
            sampleCount = 0;
            frameCount = 0;
            lastRms = 0;
            lastEnergyDbfs = -120;
        }
    }

    public void Invalidate(SessionAttemptID attempt)
    {
        lock (sync)
        {
            if (Equals(activeAttempt, attempt))
            {
                activeAttempt = null;
            }
        }
    }

    public bool TryRecordCallback(
        SessionAttemptID attempt,
        CaptureSignalMeasurement measurement,
        double? nowSeconds = null)
    {
        var callbackTime = nowSeconds ?? clock.NowSeconds;
        lock (sync)
        {
            if (!Equals(activeAttempt, attempt))
            {
                return false;
            }

            callbackCount++;
            sampleCount += Math.Max(0, measurement.SampleCount);
            frameCount += Math.Max(0, measurement.FrameCount);
            lastCallbackAtMonotonic = callbackTime;
            lastRms = measurement.Rms;
            lastEnergyDbfs = measurement.EnergyDbfs;
            return true;
        }
    }

    public CaptureSignalHealthSnapshot? Snapshot(
        SessionAttemptID? expectedAttempt = null,
        double? nowSeconds = null)
    {
        var currentTime = nowSeconds ?? clock.NowSeconds;
        lock (sync)
        {
            var currentAttempt = activeAttempt;
            if (currentAttempt is null
                || (expectedAttempt is not null && !Equals(expectedAttempt, currentAttempt)))
            {
                return null;
            }

            var elapsedSinceStart = Math.Max(0, currentTime - startedAtMonotonic);
            double? elapsedSinceLastCallback = lastCallbackAtMonotonic is { } callbackTime
                ? Math.Max(0, currentTime - callbackTime)
                : null;
            var state = callbackCount == 0
                ? elapsedSinceStart >= thresholds.InitialCallbackBudgetSeconds
                    ? CaptureSignalState.NoCallbacks
                    : CaptureSignalState.AwaitingCallbacks
                : elapsedSinceLastCallback >= thresholds.StallBudgetSeconds
                    ? CaptureSignalState.NoCallbacks
                    : lastEnergyDbfs <= thresholds.SilenceEnergyDbfs
                        ? CaptureSignalState.Silent
                        : CaptureSignalState.Audible;

            return new CaptureSignalHealthSnapshot(
                currentAttempt,
                state,
                startedAtMonotonic,
                lastCallbackAtMonotonic,
                callbackCount,
                sampleCount,
                frameCount,
                lastRms,
                lastEnergyDbfs,
                elapsedSinceStart,
                elapsedSinceLastCallback);
        }
    }
}
